src/api/software/asymmetric/ecc/soft_ecdsa.c

```diff
@@ -0,0 +1,374 @@
+        uint32_t z[curve_params->curve_wsize] __attribute__((aligned(8)));
+        uint32_t u1[curve_params->curve_wsize] __attribute__((aligned(8)));
+        uint32_t u2[curve_params->curve_wsize] __attribute__((aligned(8)));
+
+        uint32_t x1[curve_params->curve_wsize] __attribute__((aligned(8)));
+        uint32_t y1[curve_params->curve_wsize] __attribute__((aligned(8)));
+        uint32_t z1[curve_params->curve_wsize] __attribute__((aligned(8)));
+
+        /* Public key components */
+        uint32_t xq[curve_params->curve_wsize] __attribute__((aligned(8)));
+        uint32_t yq[curve_params->curve_wsize] __attribute__((aligned(8)));
+
+        ecc_bignum_affine_point_t point_aff;
+        ecc_bignum_jacobian_point_t point_jac;
+
+        /** algo 3.48 in GtECC (Guide to Elliptic Curve Cryptography) with w=2,
+         * so 2^w=4, so i=0..3, j=0..3 the array for storing the precomputed
+         * values is 16-point large */
+#define SCL_ECDSA_WINDOW_WIDTH 2
+#define SCL_ECDSA_ARRAY_SIZE                                                   \
+    (1 << SCL_ECDSA_WINDOW_WIDTH) * (1 << SCL_ECDSA_WINDOW_WIDTH)
+
+        /**
+         * variables that contain the precomputed values
+         * xP + jQ window method matrix
+         */
+        ecc_bignum_jacobian_point_t ip_jq[SCL_ECDSA_ARRAY_SIZE];
+
+        uint64_t ip_jq_x[SCL_ECDSA_ARRAY_SIZE][nb_64_bits_words_curve];
+        uint64_t ip_jq_y[SCL_ECDSA_ARRAY_SIZE][nb_64_bits_words_curve];
+        uint64_t ip_jq_z[SCL_ECDSA_ARRAY_SIZE][nb_64_bits_words_curve];
+
+        /* Copy-swap signature */
+        memset(r, 0, curve_params->curve_wsize * sizeof(uint32_t));
+        memset(s, 0, curve_params->curve_wsize * sizeof(uint32_t));
+        copy_swap_array((uint8_t *)r, signature->x, curve_params->curve_bsize);
+        copy_swap_array((uint8_t *)s, signature->y, curve_params->curve_bsize);
+
+        /* Check that r and s are in the interval [1, n-1] */
+        result = scl->bignum_func.compare(scl, (uint64_t *)r, curve_params->n,
+                                          curve_params->curve_wsize);
+        if (result >= 0)
+        {
+            result = SCL_ERROR;
+            goto cleanup;
+        }
+
+        result = scl->bignum_func.compare(scl, (uint64_t *)s, curve_params->n,
+                                          curve_params->curve_wsize);
+        if (result >= 0)
+        {
+            result = SCL_ERROR;
+            goto cleanup;
+        }
+
+        result = scl->bignum_func.is_null(scl, r, curve_params->curve_wsize);
+        if (false != result)
+        {
+            result = SCL_ERROR;
+            goto cleanup;
+        }
+
+        result = scl->bignum_func.is_null(scl, s, curve_params->curve_wsize);
+        if (false != result)
+        {
+            result = SCL_ERROR;
+            goto cleanup;
+        }
+
+        /* Copy hash into e */
+        memset(e, 0, curve_params->curve_wsize * sizeof(uint32_t));
```

Comment:
If the hash is shorter than the curve, there is a security warning. The strength is decrease to the minimum strength of either the hash or the signature algorithm. And this is check on the upper level in scl_ecdsa.c

Reply:
where is the security warning ?

Reply:
In NIST 186-4:
"A hash function that provides a lower security strength than the security strength associated with the bit length of n ordinarily should not be used, since this would reduce the security strength of the digital signature process to a level no greater than that provided by the hash function."
More over this is check in the upper layer scl_ecdsa_verification() and scl_ecdsa_signature().

Reply:
I know the document but I did not see the warning in the code

Reply:
I mean the warning in the FIPS documentation.

Reply:
In fact, after re-thinking, this is the perfect case where I should put a warning in the doc. On it.

Reply:
Done

Reply:
could you point me to the code where there is a check ?
It's not strictly prohibited by the doc but this is something important and a warning, not only in the doc, is important: it was required by a customer passing the FIPS 140.

Reply:
In scl_ecdsa.c:57 and scl_ecdsa.c:110

Reply:
ok